### PR TITLE
Ocarina of Time 7.0

### PR DESCRIPTION
**Pull request by espeon65536**
## What is this fixing or adding?
- Adds the majority of OoTR 7.0 features:
  - Pot shuffle, Freestanding item shuffle, Crate shuffle, Beehive shuffle
  - Key rings mode
  - Dungeon shortcuts to speed up dungeons
  - "Regional" shuffle for dungeon items
  - New options for shop pricing in shopsanity
  - Expanded Ganon's Boss Key shuffle options
  - Pre-planted beans
  - Improved Chest Appearance Matches Contents mode
  - Blue Fire Arrows
  - Bonk self-damage
  - Finer control over MQ dungeons and spawn position randomization
- Several bugfixes as a result of the update:
  - Items recognized by the server and valid starting items are now in a 1-to-1 correspondence. In particular, starting with keys is now supported.
  - Entrance randomization success rate improved. Hopefully it is now at 100%. 

## How was this tested?
Played part of a max-shuffle seed myself. 200 full ER generations with no failures.
Rivers' testing of various smaller options helped catch bugs with bean shuffle, triforce hunt, blue fire arrows, and shuffle child trade item.
Community playtest on 2022-12-03 (results pending).

### CODE REVIEW

File: WebHostLib/downloads.py [hunk 1]
```diff
@@ -72,7 +72,10 @@ def download_slot_file(room_id, player_id: int):
                     if name.endswith("info.json"):
                         fname = name.rsplit("/", 1)[0] + ".zip"
         elif slot_data.game == "Ocarina of Time":
-            fname = f"AP_{app.jinja_env.filters['suuid'](room_id)}_P{slot_data.player_id}_{slot_data.player_name}.apz5"
+            with zipfile.ZipFile(io.BytesIO(slot_data.data)) as zf:
```
Comment (Berserker66) [inline WebHostLib/downloads.py:75]: This looks like it will crash with a non-descript error on trying to upload an older patch?

Comment (Berserker66) [reply WebHostLib/downloads.py:75]: zipfile.is_zipfile might help, assuming the old format is not a valid zip file

Comment (espeon65536) [reply WebHostLib/downloads.py:75]: They are not, so I will make that change.

File: WebHostLib/downloads.py [hunk 2]
```diff
@@ -72,7 +72,14 @@ def download_slot_file(room_id, player_id: int):
                     if name.endswith("info.json"):
                         fname = name.rsplit("/", 1)[0] + ".zip"
         elif slot_data.game == "Ocarina of Time":
-            fname = f"AP_{app.jinja_env.filters['suuid'](room_id)}_P{slot_data.player_id}_{slot_data.player_name}.apz5"
+            stream = io.BytesIO(slot_data.data)
+            if zipfile.is_zipfile(stream):
+                with zipfile.ZipFile(stream) as zf:
```
Comment (Berserker66) [inline WebHostLib/downloads.py:77]: did you test this? we currently file.seek(0) in upload.py line 128 as is_zipfile can leave the data stream partially read. I'd be surpised if that doesn't happen here.

Comment (espeon65536) [reply WebHostLib/downloads.py:77]: I did test it and it seemed to work fine (the game patched and played properly)

Comment (Berserker66) [reply WebHostLib/downloads.py:77]: interesting. well, let's see if it breaks for anyone.

**Review by Berserker66 — APPROVED** [review]

File: WebHostLib/downloads.py (hunk 2 above)
Comment (espeon65536) [reply WebHostLib/downloads.py:77]: maybe I should just seek to 0 or make a new stream to be safe though

**Pull request merged**